File: payment-service/internal/messaging/publisher.go
package messaging

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

// Publisher sends messages to RabbitMQ.
type Publisher struct {
	conn    *amqp.Connection
	channel *amqp.Channel
}

// NewPublisher creates a Publisher connected to RabbitMQ.
func NewPublisher(amqpURL string) (*Publisher, error) {
	conn, err := connectWithRetry(amqpURL, 5)
	if err != nil {
		return nil, fmt.Errorf("publisher connect: %w", err)
	}

	ch, err := conn.Channel()
	if err != nil {
		conn.Close()
		return nil, fmt.Errorf("open channel: %w", err)
	}

	if err := declareTopology(ch); err != nil {
		ch.Close()
		conn.Close()
		return nil, fmt.Errorf("declare topology: %w", err)
	}

	return &Publisher{conn: conn, channel: ch}, nil
}

// PublishReply publishes a reply message to the saga.replies exchange.
func (p *Publisher) PublishReply(ctx context.Context, exchange, routingKey string, message interface{}) error {
	body, err := json.Marshal(message)
	if err != nil {
		return fmt.Errorf("marshal message: %w", err)
	}

	err = p.channel.PublishWithContext(
		ctx,
		exchange,
		routingKey,
		false, // mandatory
		false, // immediate
		amqp.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp.Persistent,
			Timestamp:    time.Now(),
			Body:         body,
		},
	)
	if err != nil {
		return fmt.Errorf("publish: %w", err)
	}

	log.Printf("[Publisher] Published to exchange=%s routing_key=%s", exchange, routingKey)
	return nil
}

// Close closes the channel and connection.
func (p *Publisher) Close() {
	if p.channel != nil {
		p.channel.Close()
	}
	if p.conn != nil {
		p.conn.Close()
	}
}

File: notification-service/src/messaging/publisher.js
'use strict';

const amqp   = require('amqplib');
const config = require('../config');
const logger = require('../logger');

let channel = null;

/**
 * Get or create a RabbitMQ channel for publishing.
 */
async function getChannel() {
  if (channel) return channel;

  const conn = await amqp.connect(config.rabbitmq.url);
  channel    = await conn.createChannel();

  await channel.assertExchange(config.rabbitmq.exchanges.commands, 'direct', { durable: true });
  await channel.assertExchange(config.rabbitmq.exchanges.replies,  'direct', { durable: true });

  const queues = [
    { name: 'reserve-inventory',  exchange: config.rabbitmq.exchanges.commands },
    { name: 'release-inventory',  exchange: config.rabbitmq.exchanges.commands },
    { name: 'process-payment',    exchange: config.rabbitmq.exchanges.commands },
    { name: 'refund-payment',     exchange: config.rabbitmq.exchanges.commands },
    { name: 'send-notification',  exchange: config.rabbitmq.exchanges.commands },
    { name: 'saga-replies',       exchange: config.rabbitmq.exchanges.replies   },
  ];

  for (const q of queues) {
    await channel.assertQueue(q.name, { durable: true });
    await channel.bindQueue(q.name, q.exchange, q.name);
  }

  conn.on('error',  err => { logger.error('[Publisher] Connection error', { error: err.message }); channel = null; });
  conn.on('close',  ()  => { logger.warn('[Publisher] Connection closed'); channel = null; });

  logger.info('[Publisher] RabbitMQ channel ready');
  return channel;
}

/**
 * Publish a reply message to the saga.replies exchange.
 */
async function publishReply(exchange, routingKey, message) {
  const ch   = await getChannel();
  const body = Buffer.from(JSON.stringify(message));

  ch.publish(exchange, routingKey, body, {
    persistent:   true,
    contentType:  'application/json',
    timestamp:    Math.floor(Date.now() / 1000),
  });

  logger.debug('[Publisher] Message published', { exchange, routingKey, sagaId: message.saga_id });
}

module.exports = { publishReply };
